openmc/kinetics/aux_functions.py

```diff
@@ -0,0 +1,218 @@
+    view = array[0:from_shape[0]:ratios[0],:,:,:,8:10]
+    vs = view.shape
+    view.shape = (vs[0], to_shape[1], ratios[1], to_shape[2], ratios[2],vs[3],vs[4])
+    to_array[...,8:10] = np.sum(view, axis=(2,4))
+
+    # Sum z-max currents
+    view = array[ratios[0]-1:from_shape[0]:ratios[0],:,:,:,10:12]
+    vs = view.shape
+    view.shape = (vs[0], to_shape[1], ratios[1], to_shape[2], ratios[2],vs[3],vs[4])
+    to_array[...,10:12] = np.sum(view, axis=(2,4))
+
+    to_array.shape = to_shape
+    return to_array
+
+
+def block_view(A, block):
+    """Provide a block view to 3+D array. No error checking made.
+    Therefore meaningful (as implemented) only for blocks strictly
+    compatible with the shape of A."""
+
+    block_dim = len(block)
+
+    # Check that blocks fit nicely into first dimensions of A
+    if block_dim > len(A.shape):
+        msg = 'Cannot access block_view of {}-D array with block size {}-D'\
+              ' that is greater than array dimensions'.format(len(A.shape),
+                                                              block_dim)
+        raise ValueError(msg)
+    else:
+        block = block + A.shape[block_dim:]
+        block_dim = len(A.shape)
+
+    for i,b in enumerate(block):
+        if A.shape[i] % b != 0:
+            msg = 'Cannot access block_view of {}-D array with block {}'\
+                  ' since block dimension {} does not fit evenly in'\
+                  ' Array dimension {}'.format(len(A.shape), block,
+                                               b, A.shape[i])
+            raise ValueError(msg)
+
+    # simple shape and strides computations may seem at first strange
+    # unless one is able to recognize the 'tuple additions' involved
+    shape = ()
+    strides = ()
+
+    for i,b in enumerate(block):
+        shape = shape + (int(A.shape[i] / block[i]),)
+        strides = strides + (b * A.strides[i],)
+
+    shape = shape + block
+    strides = strides + A.strides
+
+    return ast(A, shape=shape, strides=strides)
+
+def nan_inf_to_zero(array):
+
+    array[array == -np.inf] = 0.
+    array[array ==  np.inf] = 0.
+    return np.nan_to_num(array)
+
+def compute_eigenvalue(A, M, flux, tolerance=1.e-6):
+
+    # Ensure flux is a 1D array
+    flux = flux.flatten()
+
+    # Compute the initial source
+    old_source = M * flux
+    norm = old_source.mean()
+    old_source  = old_source / norm
+    flux  = flux / norm
+    k_eff = 1.0
+
+    print('initial k {}'.format((M*flux).sum() / (A*flux).sum()))
+
+    for i in range(10000):
+
+        # Solve linear system
+        flux = spsolve(A, old_source)
+
+        # Compute new source
+        new_source = M * flux
+
+        # Compute and set k-eff
+        k_eff = new_source.mean()
+
+        # Scale the new source by 1 / k-eff
+        new_source  = new_source / k_eff
+
+        # Compute the residual
+        residual_array = (new_source - old_source) / new_source
+        residual_array[residual_array == -np.inf] = 0.
```

Comment:
Use the `nan_inf_to_zero` function you had written above instead. Also, it may be good to just set this up so that infinities and nans never show up in the first place. You could instead do:

```
nonzero = new_source != 0
residual_array = (new_source[nonzero]-old_source[nonzero])/new_source[nonzero]
```

Reply:
That worked great, so included it! Thanks.